lib/lower/ASTBuilder.cpp

```diff
@@ -0,0 +1,783 @@
+  }
+
+  // When we enter a function-like structure, we want to build a new unit and
+  // set the builder's cursors to point to it.
+  template <typename A>
+  bool enterFunc(const A &f) {
+    auto &unit = addFunc(AST::FunctionLikeUnit{f, parents.back()});
+    funclist = &unit.funcs;
+    pushEval(&unit.evals);
+    parents.emplace_back(&unit);
+    return true;
+  }
+
+  void exitFunc() {
+    popEval();
+    funclist = nullptr;
+    parents.pop_back();
+  }
+
+  // When we enter a construct structure, we want to build a new construct and
+  // set the builder's evaluation cursor to point to it.
+  template <typename A>
+  bool enterConstruct(const A &c) {
+    auto &con = addEval(AST::Evaluation{c, parents.back()});
+    con.subs = new std::list<AST::Evaluation>();
+    pushEval(con.subs);
+    parents.emplace_back(&con);
+    return true;
+  }
+
+  void exitConstruct() {
+    popEval();
+    parents.pop_back();
+  }
+
+  // When we enter a module structure, we want to build a new module and
+  // set the builder's function cursor to point to it.
+  template <typename A>
+  bool enterModule(const A &f) {
+    auto &unit = addUnit(AST::ModuleLikeUnit{f, parents.back()});
+    funclist = &unit.funcs;
+    parents.emplace_back(&unit);
+    return true;
+  }
+
+  void exitModule() {
+    funclist = nullptr;
+    parents.pop_back();
+  }
+
+  template <typename A>
+  A &addUnit(const A &unit) {
+    pgm->getUnits().emplace_back(unit);
+    return std::get<A>(pgm->getUnits().back());
+  }
+
+  template <typename A>
+  A &addFunc(const A &func) {
+    if (funclist) {
+      funclist->emplace_back(func);
+      return funclist->back();
+    }
+    return addUnit(func);
+  }
+
+  /// move the Evaluation to the end of the current list
+  AST::Evaluation &addEval(AST::Evaluation &&eval) {
+    assert(funclist && "not in a function");
+    assert(evallist.size() > 0);
+    evallist.back()->emplace_back(std::move(eval));
+    return evallist.back()->back();
+  }
+
+  /// push a new list on the stack of Evaluation lists
+  void pushEval(std::list<AST::Evaluation> *eval) {
+    assert(funclist && "not in a function");
+    assert(eval && eval->empty() && "evaluation list isn't correct");
+    evallist.emplace_back(eval);
+  }
+
+  /// pop the current list and return to the last Evaluation list
+  void popEval() {
+    assert(funclist && "not in a function");
+    evallist.pop_back();
+  }
+
+  AST::Program *pgm;
+  std::list<AST::FunctionLikeUnit> *funclist{nullptr};
+  std::vector<std::list<AST::Evaluation> *> evallist;
+  std::vector<AST::ParentType> parents;
+};
+
+template <typename A>
+constexpr bool hasErrLabel(const A &stmt) {
+  if constexpr (std::is_same_v<A, Pa::ReadStmt> ||
+                std::is_same_v<A, Pa::WriteStmt>) {
+    for (const auto &control : stmt.controls) {
+      if (std::holds_alternative<Pa::ErrLabel>(control.u)) {
+        return true;
+      }
+    }
```

Comment:
nit suggestion: 
```
return std::any_of(std::begin(stmt.controls), std::end(stmt.controls), 
  [](const auto& control) { return std::holds_alternative<Pa::ErrLabel>(control.u); });
```

Reply:
Or even better because this happens multiple times, add:
```
auto isError = [](const auto& v) { return std::holds_alternative<Pa::ErrLabel>(v.u); });
```
before the if constexpr chain and then use
```
return std::any_of(std::begin(stmt.controls), std::end(stmt.controls), isError);
```
in each if constexpr branch.

Reply:
Thanks for the suggestion, it seemed to me a good idea so I did do that.
@klausler, what did you down-voted that ?

Reply:
The range-based `for` idiom is more clear than the cluttered call to the iterator-based "algorithm".

Reply:
I don't really agree that the range-for is more clear, I think in the any_of case it's really easy to see that you're trying to do a boolean check on the entire vector. However as I said in my suggestion I don't feel particularly strongly about it, I just thought this read a little better.